### DISCUSSION TITLE
Jest/supertest hanging on async operations not stopped in tests

### DISCUSSION THREAD

**Discussion by jameynakama** [prisma/prisma#5232] [Q&A] [answered]
Hey, everybody. I'm trying to write tests for a simple endpoint that fetches a record, and I can't figure out how to get around this error:

```
Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```

I've got this route:

```javascript
app.get('logs/:id', async (req, res) => {
  const log = await prisma.ingestionLog.findUnique({
    where: {
      id: parseInt(req.params.id)
    }
  });
  res.json(log);
});
```

And a test:

```javascript
    it('should return one record', async done => {
      await prisma.ingestionLog.create({
        data: {
          message: "test-thing"
        }
      });

      request(app).get('logs/1')
        .expect(200)
        .end((err, res) => {
          if (err) return done(err);
          expect(res.body.message).toBe('test-thing');
          done();
        });
    });
```

The test passes, but hangs on that error. Adding `--detectOpenHandles` doesn't provide any additional output.

If I change the endpoint just to send some text, the tests exit gracefully. It's only when the endpoint uses prisma to grab records that the tests hang.

I've got this hook:

```javascript
afterAll(async () => {
  await prisma.$disconnect();
  app.close();
});
```

...which doesn't help.

**Comment by jameynakama** [comment]
I've found that adding a `disconnect` inside the routes allows the tests to finish gracefully, but that seems like the wrong approach?

```js
app.get(appRoute('logs'), async (req, res) => {
  const logs = await prisma.ingestionLog.findMany();
  await prisma.$disconnect();
  return res.json(logs)
});
```

**Comment by jameynakama** [comment, marked as answer]
Solved this. I noticed I was creating a new prisma client in my test files instead of importing the module-level one created by the app. I switched to that and now only need the `await prisma.$disconnect();` in `afterAll` and everything's running smoothly.

I got rid of the `app.close()` line by separating my express app creation and server creation (the call to `app.listen(...)`), and only importing the raw app in my tests.

**Reply by jrbytes** [reply]
My tests seem to be ok, but when I check with --detectOpenHandles it returns some messages.
My app file is not starting the app with app listen.

<img width="876" alt="Screen Shot 2022-09-22 at 18 39 45" src="https://user-images.githubusercontent.com/26124887/191856265-f3998ac4-cef0-400a-980f-64806ab0893f.png">

File test
```javascript
import request from 'supertest'

import { prisma } from '@database/prisma'
import { app } from '@infra/http'

describe('Create Client Controller', () => {
  beforeAll(async () => {
    await prisma.clients.create({
      data: {
        username: 'username',
        password: 'password'
      }
    })
  })

  afterAll(async () => {
    const deleteClient = prisma.clients.deleteMany()
    await prisma.$transaction([deleteClient])
    await prisma.$disconnect()
  })

  it('should be able to create an client', async () => {
    const responseClient = await request(app).post('/clients').send({
      username: 'jrbytes',
      password: 'password'
    })

    expect(responseClient.status).toBe(201)
  })

  it('should not be able to create an client if has exists', async () => {
    const responseClient = await request(app).post('/clients').send({
      username: 'username',
      password: 'password'
    })

    expect(responseClient.status).toBe(400)
    expect(responseClient.body).toHaveProperty('message')
    expect(responseClient.body.message).toBe('Client already exists')
  })
})
```

File prismaClient
```javascript
import { PrismaClient } from '@prisma/client'

export const prisma = new PrismaClient({
  datasources: {
    db: {
      url: process.env.NODE_ENV === 'test' ? process.env.TEST_DATABASE_URL : process.env.DATABASE_URL
    }
  }
})
```

**Comment by jrbytes** [comment]
My tests seem to be ok, but when I check with --detectOpenHandles it returns some messages.
My app file is not starting the app with app listen.

<img width="876" alt="Screen Shot 2022-09-22 at 18 39 45" src="https://user-images.githubusercontent.com/26124887/191856265-f3998ac4-cef0-400a-980f-64806ab0893f.png">

File test
```javascript
import request from 'supertest'

import { prisma } from '@database/prisma'
import { app } from '@infra/http'

describe('Create Client Controller', () => {
  beforeAll(async () => {
    await prisma.clients.create({
      data: {
        username: 'username',
        password: 'password'
      }
    })
  })

  afterAll(async () => {
    const deleteClient = prisma.clients.deleteMany()
    await prisma.$transaction([deleteClient])
    await prisma.$disconnect()
  })

  it('should be able to create an client', async () => {
    const responseClient = await request(app).post('/clients').send({
      username: 'jrbytes',
      password: 'password'
    })

    expect(responseClient.status).toBe(201)
  })

  it('should not be able to create an client if has exists', async () => {
    const responseClient = await request(app).post('/clients').send({
      username: 'username',
      password: 'password'
    })

    expect(responseClient.status).toBe(400)
    expect(responseClient.body).toHaveProperty('message')
    expect(responseClient.body.message).toBe('Client already exists')
  })
})
```

File prismaClient
```javascript
import { PrismaClient } from '@prisma/client'

export const prisma = new PrismaClient({
  datasources: {
    db: {
      url: process.env.NODE_ENV === 'test' ? process.env.TEST_DATABASE_URL : process.env.DATABASE_URL
    }
  }
})
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
